### ISSUE TITLE
Transplanting Report in Barn Kit Tab

### ISSUE THREAD

**Issue by braughtg** [DickinsonCollege/FarmData2#139]
See Seed -> Transplanting -> Report in the original FarmData at [https://farmdata.dickinson.edu/guest.php](https://farmdata.dickinson.edu/guest.php)

Data report for transplanting. Provides raw data of crop transplanted over selected timeframe, plus information about seeding and transplanting dates and duration.

Also see #136 for some additional general language that applies to all reports.

**Comment by IrisSC** [comment]
Contributors: IrisSC, boosungkim, and miyu386
Overall: 

The group worked on Issue #139 to create a page for generating a report for transplanting data.  This issue primarily focuses on the transplanting page under the Barn Kit subtab. There are input entries and a table, where the filtered harvest reports are displayed.  

Link to the entire Pull Request: https://github.com/DickinsonCollege/FarmData2/pull/159 

Billy’s completed stories: 

Row Feet/ Field data- This story is about displaying the row feet of the transplanting crop. This is used along with Row Feet/Bed to calculate the Bed Feet value as it is not given directly in the API. The story was very simple as there was no need to create a filter section for the users and it was very easily accessed through the API within the quantity array under ‘measure’. Now, because the data retrieved had units along with the number, it was necessary to create a method function to split the string and take only the number. This was called getRowFeet(). With the retrieved log data, it was then displayed on the table under the appropriate column.  

 

Boo Sung’s completed stories: 

Filter by Type of Crop - This story was about allowing the user to filter the harvest reports by the name of the crop. There were two main features to this story: filtering the harvest reports by the crop name and displaying the corresponding reports in the table. The value of the chosen crop name from the input field was saved in the Vue.js instance, and the URL link query was updated to display reports with that particular crop. Once the reports are retrieved, the code accesses the crop name and displays it in the table. The main difficulty was that the transplanted crop names are different from normal crop names - instead of “Basil,” it would be “Transplanting Basil.” 

Rows in Bed - This story was about displaying the number of rows in the bed of the harvest in the table. This was pretty simple, as the code only had to access the rows in bed under the “quantity” array in the API. Professor Braught and we agreed that filtering by the number of rows in bed is not necessary, so only the display function was implemented. 

Username - This story was about displaying the username of the user who made harvest report in the table. The code accessed the API using the specified URL query, and retrieved and stored the username in the Vue.js instance. Then, it was returned to the table, where it is displayed. 

Iris’ completed stories: 

Filter by Time Period – For this story two date inputs were created. One was labeled start date and the other was labeled end date. The value of these inputs are storied in the Vue.js instance. They are also then converted into timestamp dates that are used to select a range of dates from the start date input to the end date input. The code is set up so that the date inputs are allowed to be empty. In that case the data is from all dates or only has a start or end date parameter. It is also impossible to select a start date after an end date and vice versa. This story required both the standard dates to timestamp and the timestamp back to standard date methods. 

Transplanting Date – For this story the time was calculated form the timestamp variable in each log and was then printed on the table for each log. This story required the timestamp to standard date method to complete. There are two other timestamps in each log: created and changed. This both correct to the time that the log was create for our purposes and changed respectively. Both dates are somewhere in 2020 and most of the timestamps for the logs are in 2018. This means that these timestamps do not represent when the log was created. 

Bed Feet – For this story the rows in bed need to be calculated and then added to the table. The Bed Feet were calculated by having the row feet divide by the rows in the bed. To do this a calculate bed feet method was created that took in row feet and rows in bed. These were then run through two different methods that got the number from the rest of the string. Then the bed feet were calculated and returned, 

Comments – The comments were found in the value key of the notes array. They are in html format but when a put them into the table that code (e.i. <p>) was still in there. For now, I just took out the html sections and have it printed in the table. It might be helpful in the future to change this so that the html is converted in the table. 

 

Stories in progress: 

Field name: This story is about allowing users to filter the transplanting data based on the field name of the crops and displaying them in the table. There are two main parts to this story, enabling the feature to allow users to filter data by field name, and implementing the field name column in the data table. Implementing the field name in the table was able to be done easily through accessing the API. Very similar to other stories, I simply had to find the right parameters in the API and display that in the table. However, a problem was generated when I worked to filter the whole log by the field name, as I could not create an API endpoint link that worked, as it either did not display any data or it just generated an error. As there was no specific guideline to this, it was a trial-error process where I tried using different parameter names but none of it worked. Furthermore, I looked into alternative solutions such as finding a different parameter name that could be used to filter data by field which I wasn’t able to find, and as professor braught suggested, making the filtering process within the JavaScript code. The idea was to filter the log accessed after the API request was made and filtering it with if statements. However, it created error yet again and I(team) was not able to find a solution.  

 

Incomplete stories: 

The only stories that the team was unable to complete were features that existed in the original farm data website but does not exist in the current farm data website. 

The current incomplete stories are: Date of Seeding, Days in Tray, Number of Trays, and Hours Spent Transplanting. For all these stories, there are no corresponding variables in the log and the API. The seeding date story is likely in the seeding log, meaning that someone would need to connect the transplanting log to its seeding log. The others are unknown.
